File: leetcode/BackTracking/leetcode-140/140.go
package backtracking

func wordBreak(s string, wordDict []string) []string {
	m := make(map[string][]string)
	return helper(s, wordDict, m)
}

func helper(s string, wordDict []string, m map[string][]string) []string {
	if _, ok := m[s]; ok {
		return m[s]
	}

	res := []string{}
	if len(s) == 0 {
		res = append(res, "")
		return res
	}

	for _, word := range wordDict {
		if len(word) > len(s) {
			continue
		}

		if s[:len(word)] != word {
			continue
		}

		rem := helper(s[len(word):], wordDict, m)
		for _, str := range rem {
			var tmp string
			if len(str) == 0 {
				tmp = word
			} else {
				tmp = word + " " + str
			}
			res = append(res, tmp)
		}
	}

	m[s] = res
	return res
}

File: leetcode/BackTracking/leetcode-140/140.java
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        Map<String, List<String>> map = new HashMap<>();
        return helper(s, wordDict, map);
    }

    private List<String> helper(String s, List<String> wordDict, Map<String, List<String>> map) {
        if (map.containsKey(s)) {
            return map.get(s);
        }
        
        List<String> res = new ArrayList<>();
        if (s.length() == 0) {
            res.add("");
            return res;
        }

        for (String word : wordDict.toArray(new String[wordDict.size()])) {
            if (word.length() > s.length()) {
                continue;
            }
            if (s.substring(0, word.length()).compareTo(word) != 0) {
                continue;
            }
            List<String> rem = helper(s.substring(word.length()), wordDict, map);
            for (String str : rem.toArray(new String[rem.size()])) {
                String tmp = word + ((str.length() == 0) ? "" : " ") + str;
                res.add(tmp);
            }
        }
        map.put(s, res);
        return map.get(s);
    }
}